libflux/src/core/semantic/convert.rs

```diff
@@ -233,6 +234,59 @@ fn convert_monotype(
     }
 }
 
+#[allow(unused)]
+fn convert_polytype(
+    type_expression: ast::TypeExpression,
+    f: &mut Fresher,
+) -> Result<types::PolyType> {
+    // pub vars: Vec<Tvar>,
+    // pub cons: TvarKinds,
+    // pub expr: MonoType,
+    // ==
+    // pub base: BaseNode,
+    // pub monotype: MonoType,
+    // #[serde(skip_serializing_if = "Vec::is_empty")]
+    // pub constraints: Vec<TypeConstraint>,
+    // (a: T, b: T) => T where T: Addable + Divisible
+    //     |---------------|       |--------------------|
+    //     expr                       cons
+    //                         |  |-----|   |-------|
+    //                       var1  kind1      kind2
+    let mut tvars = HashMap::<String, types::Tvar>::new();
+    let mut expr = convert_monotype(type_expression.monotype, &mut tvars, f)?;
+    let mut vars = Vec::<types::Tvar>::new();
+    let mut cons = SemanticMap::<types::Tvar, Vec<types::Kind>>::new();
+
+    for c in type_expression.constraints {
+        if tvars.contains_key(c.tvar.name.as_str()) {
+            vars.push(match tvars.get(&c.tvar.name) {
+                Some(v) => *v,
+                None => {types::Tvar(0)} // never reached.
```

Comment:
This should return an error.
```Rust
None => Err(format!("type variable {} constrained but not used", ...))
```

May I suggest something along the lines of the following (more or less):
```Rust
for c in type_expression.constraints {
    match tvars.remove(c.tvar.name) {
        None => {
            return Err(...);
        }
        Some(tv) => {
            vars.push(tv);
        }
    }
...
```
In other words, just remove the element from the map. If `None` then throw an error.